types/hooks.d.ts

```diff
@@ -0,0 +1,235 @@
+    export type KeyframeFn<DS extends object> = (
+      next: (props: UseSpringProps<DS>) => Promise<void>,
+      cancel: () => void
+    ) => Promise<void>
+    interface SpringKeyframeSlotsConfig extends HooksBaseProps {
+      // this is way too self-referential to type correctly
+      // both onRest and config have to have vague types... 😭
+
+      /**
+       * Callback when the animation comes to a still-stand
+       */
+      onRest?(ds: any): void
+      config?:
+        | SpringConfig
+        | ReadonlyArray<SpringConfig>
+        | ((slot: string) => SpringConfig | ReadonlyArray<SpringConfig>)
+    }
+    interface TrailKeyframeSlotsConfig<TItem>
+      extends SpringKeyframeSlotsConfig {
+      items: ReadonlyArray<TItem>
+    }
+
+    type ResolveKeyframeSlotValue<T> = T extends KeyframeFn<infer V>
+      ? V
+      : T extends ReadonlyArray<infer U>
+      ? U extends KeyframeFn<infer V>
+        ? V
+        : U extends Function // guard against accidentally putting in a KeyframeFn function
+        ? never
+        : U
+      : T extends Function // same guard, but when the function is not in an array
+      ? never
+      : T
+
+    // fun fact: the state is literally named "undefined" if you're using this overload
+    // the docs are vague but it seems this just loops the one animation forever?
+    export function spring<DS extends object>(
+      animation:
+        | ReadonlyArray<DS>
+        | ReadonlyArray<KeyframeFn<DS>>
+        | KeyframeFn<DS>,
+      initialProps?: UseSpringProps<DS>
+    ): UseSpringKeyframes<DS>
+    // unfortunately, it's not possible to infer the type of the callback functions (if any are given)
+    // while also remaining possible to infer the slot names. Callback functions have to be cast with
+    // `as useKeyframes.KeyframeFn<{ ... }>`.
+    // it's also not possible to specify the types of the values inside TSlots. This is a mess.
+    export function spring<TSlots extends object>(
+      slots: TSlots & SpringKeyframeSlotsConfig,
+      // also unfortunately not possible to strongly type this either
+      initialProps?: UseSpringProps<any>
+    ): UseSpringKeyframesWithSlots<TSlots>
+
+    export function trail<DS extends object>(
+      animation:
+        | ReadonlyArray<DS>
+        | ReadonlyArray<KeyframeFn<DS>>
+        | KeyframeFn<DS>,
+      initialProps?: UseSpringProps<DS>
+    ): UseTrailKeyframes<DS>
+    export function trail<TItem, TSlots extends object>(
+      slots: TSlots & TrailKeyframeSlotsConfig<TItem>,
+      initialProps?: UseSpringProps<any>
+    ): UseTrailKeyframesWithSlots<TSlots>
+
+    type UseSpringKeyframesWithSlots<TSlots extends object> =
+      // there's a bug in the implementation that actually should cause a crash
+      // because there is no second argument, but because it is built with babel on loose mode,
+      // it doesn't...
```

Comment:
Babel is compiling `const [state, count] = ...` to array index accesses, which "work" (they return `undefined`), but if there was no loose mode babel, this would run the iterator and hit `done: true` before the expected number of iterations.

The proper fix would be to have `const [state = undefined, count = undefined] = ...`; that wouldn't crash on real engines either.